### PR TITLE
Bug 1530714 - Report ASRouter errors in ASRouterAdmin

### PR DESCRIPTION
**Pull request by piatra**
This adds a new section in ASRouterAdmin where you can see all logged errors.

image

I've broken down the `_remoteLoader` to report missing messages, response failures or failures to parse json. Also all calls to `Cu.reportError` now go through the same function that captures them to be displayed in the admin panel.

### CODE REVIEW

File: lib/ASRouter.jsm [hunk 1]
```diff
@@ -49,6 +49,16 @@ const STARTPAGE_VERSION = "6";
...
   STARTPAGE_VERSION,
   REMOTE_LOADER_CACHE_KEY: "RemoteLoaderCache",
+  _errors: [],
+
+  reportError(e) {
+    Cu.reportError(e);
+    this._errors = [{timestamp: new Date(), error: e}, ...this._errors];
+  },
+
+  get errors() {
```
Comment (k88hudson) [inline lib/ASRouter.jsm]: I like the `reportError` strategy, but personally I would return the error through `loadMessagesForProvider` so the state can be centralized in `ASRouter._state`, or maybe even as a property inside provider in `ASRouter._state.providers` itself?

Keeping it here makes cleanup more complicated IMO

File: lib/ASRouter.jsm [hunk 2]
```diff
@@ -104,17 +114,29 @@ const MessageLoaderUtils = {
...
-          (response.ok || response.status === 302)
-        ) {
-          remoteMessages = (await response.json())
-            .messages
+        response = await fetch(provider.url, {headers});
+      } catch (e) {
+        MessageLoaderUtils.reportError(e);
+      }
+      if (
+        response &&
```
Comment (k88hudson) [inline lib/ASRouter.jsm]: This might be a little too strict, personally I usually go with `status >= 200 && status < 400`

**Comment by piatra** [comment]
Updated the PR
* We now only report `remote` provider errors
* `errors` are part of the `provider` state
* reverted some of the `Cu.reportError` changes so we store provider errors and everything else is reported to the console
* reporting `provider` id now with the errors
image

File: lib/ASRouter.jsm [hunk 3]
```diff
@@ -104,17 +116,26 @@ const MessageLoaderUtils = {
...
-          remoteMessages = (await response.json())
-            .messages
+        response = await fetch(provider.url, {headers, credentials: "omit"});
+      } catch (e) {
+        MessageLoaderUtils.reportError(e);
+      }
+      if (
+        response &&
+        (response.status >= 200 && response.status < 400) &&
+        (response.ok || response.status === 302)
```
Comment (k88hudson) [inline lib/ASRouter.jsm]: You don't need `=== 302` if you have `<400`

File: content-src/components/ASRouterAdmin/ASRouterAdmin.jsx [hunk 4]
```diff
@@ -562,6 +562,37 @@ export class ASRouterAdminInner extends React.PureComponent {
       </div>);
   }
 
+  renderErrorMessage({id, errors}) {
+    const providerId = <td rowSpan={errors.length}>{id}</td>;
+    // .reverse() so that the last error (most recent) is first
+    return errors.reverse().map(({error, timestamp}, cellKey) => (<tr key={cellKey}>
```
Comment (k88hudson) [inline content-src/components/ASRouterAdmin/ASRouterAdmin.jsx]: can you `.reverse` after `.map` (don't mutate input)

**Pull request merged**